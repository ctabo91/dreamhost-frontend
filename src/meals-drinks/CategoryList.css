.CategoryCard {
    background-color: rgb(0, 0, 0);
    color: rgb(110, 66, 66);
    text-decoration: none;
    font-family: "Tenor Sans", sans-serif;
    font-weight: 400;
    font-style: normal;
}

.CategoryCard:hover {
    color: rgb(0, 0, 0);
    background-color: rgb(110, 66, 66);
    text-decoration: none;
}

.CategoryList-title {
    font-family: "Gruppo", sans-serif;
    font-weight: 400;
    font-style: normal;
    font-size: 50px;
}

