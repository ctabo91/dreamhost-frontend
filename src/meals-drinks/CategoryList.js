import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import DreamHostApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";
import CategoryCard from "./CategoryCard";
import "./CategoryList.css";

/** Shows a list of CategoryCards depending on the value of "items"
 */ 

function CategoryList() {
    const { item } = useParams();
    const [categories, setCategories] = useState(null);

    // Get categories for the specific item passed (meals or drinks).
    useEffect(() => {
        const getCategories = async (type) => {
            try {
                const categories = await DreamHostApi.getCategories(type);
                setCategories(categories);
            } catch (err) {
                console.error("Error fetching categories:", err);
            }
        };

        getCategories(item);
    }, [item]);

    if (!categories) return <LoadingSpinner />;

    return (
        <div className="CategoryList col-md-8 offset-md-2 mt-5 pt-5">
            <h4 className="CategoryList-title display-4 text-center mb-5">
                {item === "meals"
                        ? "- Select A Meal Category -"
                        : "- Select A Drink Category -"
                }
            </h4>
            {categories.length
                ? (
                    <div className="CategoryList-list row row-cols-1 row-cols-md-3">
                        {categories.map((c, idx) => (
                            <CategoryCard 
                                key={idx}
                                category={c.category}
                                count={c.count}
                                item={item}
                            />
                        ))}
                    </div>
                ) : (
                    <p className="lead">Sorry, no results were found!</p>
                )}
        </div>
    );
}

export default CategoryList;
