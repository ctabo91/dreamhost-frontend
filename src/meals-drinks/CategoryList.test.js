import React from 'react';
import { render, screen, act } from '@testing-library/react';
import CategoryList from './CategoryList';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import DreamHostApi from '../api/api';

jest.mock('../api/api');

const mockCategories = [
  { category: 'Category 1', count: 5 },
  { category: 'Category 2', count: 6 },
  { category: 'Category 3', count: 7 },
];

beforeEach(() => {
  jest.clearAllMocks();
});

describe('CategoryList Component', () => {
  it("matches snapshot", async () => {
    DreamHostApi.getCategories.mockResolvedValueOnce(mockCategories);

    const asFragment = await act(async () => {
      const { asFragment } = render(
        <MemoryRouter initialEntries={['/categories/meals']}>
          <Routes>
            <Route path="/categories/:item" element={<CategoryList />} />
          </Routes>
        </MemoryRouter>
      );
      return asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
  })

  it('renders loading spinner while fetching categories', async () => {
    await act(async () => {
      render(
        <MemoryRouter initialEntries={['/categories/meals']}>
          <Routes>
            <Route path="/categories/:item" element={<CategoryList />} />
          </Routes>
        </MemoryRouter>
      );
    });

    expect(screen.getByText('Loading ...')).toBeInTheDocument();
  });

  it('renders meal categories correctly', async () => {
    DreamHostApi.getCategories.mockResolvedValueOnce(mockCategories);

    await act(async () => {
      render(
        <MemoryRouter initialEntries={['/categories/meals']}>
          <Routes>
            <Route path="/categories/:item" element={<CategoryList />} />
          </Routes>
        </MemoryRouter>
      );
    });

    // Wait for the API call to resolve
    await screen.findByText('- Select A Meal Category -');

    expect(screen.getByText('CATEGORY 1')).toBeInTheDocument();
    expect(screen.getByText('CATEGORY 2')).toBeInTheDocument();
    expect(screen.getByText('CATEGORY 3')).toBeInTheDocument();
  });

  it('renders drink categories correctly', async () => {
    DreamHostApi.getCategories.mockResolvedValueOnce(mockCategories);

    await act(async () => {
      render(
        <MemoryRouter initialEntries={['/categories/drinks']}>
          <Routes>
            <Route path="/categories/:item" element={<CategoryList />} />
          </Routes>
        </MemoryRouter>
      );
    });

    // Wait for the API call to resolve
    await screen.findByText('- Select A Drink Category -');

    expect(screen.getByText('CATEGORY 1')).toBeInTheDocument();
    expect(screen.getByText('CATEGORY 2')).toBeInTheDocument();
    expect(screen.getByText('CATEGORY 3')).toBeInTheDocument();
  });

  it('displays "Sorry, no results were found!" when no categories are available', async () => {
    DreamHostApi.getCategories.mockResolvedValueOnce([]);

    await act(async () => {
      render(
        <MemoryRouter initialEntries={['/categories/meals']}>
          <Routes>
            <Route path="/categories/:item" element={<CategoryList />} />
          </Routes>
        </MemoryRouter>
      );
    });

    // Wait for the API call to resolve
    await screen.findByText('Sorry, no results were found!');

    expect(screen.getByText('Sorry, no results were found!')).toBeInTheDocument();
  });
});
