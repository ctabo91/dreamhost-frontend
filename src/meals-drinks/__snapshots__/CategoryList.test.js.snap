// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CategoryList col-md-8 offset-md-2 mt-5 pt-5"
  >
    <h4
      class="CategoryList-title display-4 text-center mb-5"
    >
      - Select A Meal Category -
    </h4>
    <div
      class="CategoryList-list row row-cols-1 row-cols-md-3"
    >
      <div
        class="col-4 mb-4"
      >
        <a
          class="CategoryCard card border-0"
          href="/recipes/meals/Category%201"
        >
          <div
            class="card-body text-center pl-0 pr-0"
          >
            <h5
              class="card-title"
            >
              CATEGORY 1
            </h5>
            <p
              class="card-text"
            >
              <small>
                5 recipes
              </small>
            </p>
          </div>
        </a>
      </div>
      <div
        class="col-4 mb-4"
      >
        <a
          class="CategoryCard card border-0"
          href="/recipes/meals/Category%202"
        >
          <div
            class="card-body text-center pl-0 pr-0"
          >
            <h5
              class="card-title"
            >
              CATEGORY 2
            </h5>
            <p
              class="card-text"
            >
              <small>
                6 recipes
              </small>
            </p>
          </div>
        </a>
      </div>
      <div
        class="col-4 mb-4"
      >
        <a
          class="CategoryCard card border-0"
          href="/recipes/meals/Category%203"
        >
          <div
            class="card-body text-center pl-0 pr-0"
          >
            <h5
              class="card-title"
            >
              CATEGORY 3
            </h5>
            <p
              class="card-text"
            >
              <small>
                7 recipes
              </small>
            </p>
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
